### PR TITLE
Created functions to generate random k-tree and partial k-tree graphs

### PR DESCRIPTION
**Pull request by damianbasso**
Implemented graph generators which produce a random k-tree and a random partial k-tree. The k-tree algorithm first generates a complete graph, before adding vertices one by one, creating a clique utilising random existing vertices and the new vertex. Partial k-trees are generated by producing a k-tree and then randomly removing edges from it.

Previously SageMath lacked a way to generat randomised graphs which have a fixed treewidth value. This provides a methodology to do so .

Addresses #36586

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by dcoudert** [comment]
For your information, there is an (old) open issue about generating k-trees #11369.

File: src/sage/graphs/generators/random.py [hunk 1]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
         return False
 
 
+def RandomKTree(n, k, seed=None):
+    r"""
+    Returns a random k-tree on `n` nodes numbered `0` through `n-1`.
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: `Returns` -> `Return`
k-tree -> ``` `k` ```-tree

File: src/sage/graphs/generators/random.py [hunk 2]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+
+    ALGORITHM:
+
+    The algorithm first generates a complete graph on `k+1` vertices. Vertices are
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: make sure than comments are in 80 columns mode

File: src/sage/graphs/generators/random.py [hunk 3]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+    subsequently generated by randomly choosing one of the existing cliques in the graph,
+    and creating a new clique by replacing one of the vertices in the selected clique with 
+    a newly created one.
+
+    INPUT:
+
+    -  ``n`` -- number of vertices in the k-tree
+
+    -  ``k`` -- within a clique each vertex is connected to `k` vertices. `k` also corresponds
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: also here, use 80 columns mode.

File: src/sage/graphs/generators/random.py [hunk 4]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+      to the treewidth of the k-tree
+
+    - ``seed`` -- a ``random.Random`` seed or a Python ``int`` for the random
+      number generator (default: ``None``)
+
+    TESTS::
+        sage: g=graphs.RandomKTree(50,5)
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py:1485]: add an empty line after `TESTS::`

File: src/sage/graphs/generators/random.py [hunk 5]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        sage: g.size()
+        235
+        sage: len(g.vertices(sort=False))
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: you can simply use `g.order()`

File: src/sage/graphs/generators/random.py [hunk 6]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        50
+        sage: g.treewidth()
+        5
+
+    EXAMPLES::
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: To document that random k-trees have tree width k, you may do:
```
    EXAMPLES:

    A (random) `k`-tree has tree width `k`::

        sage: G = graphs.RandomKTree(50, 5)
        sage: G.treewidth()
        5
```

File: src/sage/graphs/generators/random.py [hunk 7]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+
+        sage: G = graphs.RandomKTree(50,5)
+        sage: G.treewidth()
+        5
+        sage: G.show()
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: please change to: `sage: G.show()  # not tested`.
This is to avoid plotting the graph when testing the module. Plotting is nice for the documentation only.

File: src/sage/graphs/generators/random.py [hunk 8]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+
+    EXAMPLES::
+
+        sage: G = graphs.RandomKTree(50,5)
+        sage: G.treewidth()
+        5
+        sage: G.show()      
+    """
+
+    from sage.graphs.generators.basic import CompleteGraph
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: the import should be done after testing the validity of the parameters. Also, there is no need for the empty line before.

File: src/sage/graphs/generators/random.py [hunk 9]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+    
+    if (n<k+1):
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: `if n < k + 1:`

File: src/sage/graphs/generators/random.py [hunk 10]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        raise ValueError("n must be greater than k + 1.")
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: avoid the final `.`

File: src/sage/graphs/generators/random.py [hunk 11]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py:1531]: add
```
    if seed is not None:
        set_random_seed(seed)
```

File: src/sage/graphs/generators/random.py [hunk 12]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+    g = CompleteGraph(k+1)
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: For your information, you could do:
```
G = Graph()
G.add_clique(list(range(k + 1)))
```

File: src/sage/graphs/generators/random.py [hunk 13]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+    g.name("")
+    
+    cliques = [list(range(k+1))]
+
+    # Randomly choose a row, and copy 1 of the cliques
+    # One of those vertices is then replaced with a new vertex 
+    for i in range(n-k-1):
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: here, you can simply use:
```py
    for newVertex in range(k + 1, n):
```

File: src/sage/graphs/generators/random.py [hunk 14]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        cliques.append(copiedClique)
+        for u in copiedClique:
+            if u != newVertex:
+                g.add_edge(u, newVertex)
+    return g    
+
+
+def RandomPartialKTree(n, k, x, seed=None):
+    r"""
+    Forms a random k-tree on `n` nodes, and then removes `x` random edges from it leaving
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: Change to
```
    Return a random partial `k`-tree on `n` nodes.

    A partial `k`-tree is defined as....
```

and make sure to use 80 columns mode

File: src/sage/graphs/generators/random.py [hunk 15]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        230
+        sage: g.treewidth()
+        5
+
+    EXAMPLES:
+
+        sage: G = graphs.RandomKTree(50,5)
+        sage: G.treewidth()
+        5
+        sage: G.show(
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: missing ) and please add ` # not tested`

File: src/sage/graphs/generators/random.py [hunk 16]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+    a partial k-tree.
+
+    INPUT:
+
+    -  ``n`` -- number of vertices in the k-tree
+
+    -  ``k`` -- within a clique each vertex is connected to `k` vertices. `k` also corresponds
+      to the treewidth of the k-tree
+
+    -  ``x`` -- how many vertices are deleted from the k-tree
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: you remove edges, right ?

File: src/sage/graphs/generators/random.py [hunk 17]
```diff
@@ -1460,6 +1460,116 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        sage: G.show(
+    """
+    if seed is not None:
+        set_random_seed(seed)
+
+    g = RandomKTree(n,k,seed)
+
+    # Check that x doesn't delete too many edges
+    if x > g.size():
+        raise ValueError("x must be less than the number of nodes in the K-Tree with `n` nodes.")
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: number of nodes or edges ? you compared `x` to the number of edges of `g`.  Also, avoid the final `.`

**Review by dcoudert — CHANGES_REQUESTED** [review]
Thank you for this contribution. Random k-trees are useful.

I have a few suggestions for improving the code.

File: src/sage/graphs/generators/random.py (hunk 17 above)
Comment (damianbasso) [reply src/sage/graphs/generators/random.py]: Yep, should be edges

**Comment by damianbasso** [comment]
> Thank you for this contribution. Random k-trees are useful.
> 
> I have a few suggestions for improving the code.

Thanks for reviewing @dcoudert. Just added another commit implementing your suggestions.

File: src/sage/graphs/generators/random.py [hunk 18]
```diff
@@ -1460,6 +1460,123 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        sage: g.size()
+        235
+        sage: g.order()
+        50
+        sage: g.treewidth()
+        5
+
+    EXAMPLES:
+
+    A (random) `k`-tree has tree width `k`::
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: `treewidth`

File: src/sage/graphs/generators/random.py [hunk 19]
```diff
@@ -1460,6 +1460,123 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+    """    
+    if n < k + 1:
+        raise ValueError("n must be greater than k + 1")
+
+    if seed is not None:
+        set_random_seed(seed)
+
+    from sage.graphs.generators.basic import CompleteGraph
+
+    g = Graph()
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: may be you can set the name of the graph like `g = Graph(name=f"Random {k}-tree of order {n}")` ?

File: src/sage/graphs/generators/random.py [hunk 20]
```diff
@@ -1460,6 +1460,123 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+                g.add_edge(u, newVertex)
+    return g 
+
+
+def RandomPartialKTree(n, k, x, seed=None):
+    r"""
+    Return a random partial `k`-tree on `n` nodes. 
+
+    A partial `k`-tree is defined as a subgraph of a `k`-tree. This can also be
+    described as a graph with treewidth at most k.
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: ``` `k` ```

same in several other places in both methods.

File: src/sage/graphs/generators/random.py [hunk 21]
```diff
@@ -1460,6 +1460,123 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        sage: g.order()
+        50
+        sage: g.size()
+        230
+        sage: g.treewidth()
+        5
+
+    EXAMPLES:
+
+        sage: G = graphs.RandomKTree(50,5)
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: `RandomPartialKTree` ?

File: src/sage/graphs/generators/random.py [hunk 22]
```diff
@@ -1460,6 +1460,123 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        5
+        sage: G.show()
+    """
+    if seed is not None:
+        set_random_seed(seed)
+
+    g = RandomKTree(n,k,seed)
+
+    # Check that x doesn't delete too many edges
+    if x > g.size():
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: there is a formula for the number of edges of a k-tree, so you can compare x to that number before building the k-tree.

File: src/sage/graphs/generators/random.py [hunk 23]
```diff
@@ -1460,6 +1460,123 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        raise ValueError("x must be less than the number of edges in the K-Tree with `n` nodes")
+    
+    for i in range(x):
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: a faster alternative is to get the list of edges, then shuffle it (we have a method to shuffle a list) and finally remove the x first edges of that list from the graph.

**Review by dcoudert — CHANGES_REQUESTED** [review]
a few more suggestions.

**Comment by damianbasso** [comment]
> a few more suggestions.

Thanks, some good ideas. I've added to the Pull Request

File: src/sage/graphs/generators/random.py [hunk 24]
```diff
@@ -1567,13 +1565,16 @@ def RandomPartialKTree(n, k, x, seed=None):
...
 
     # Check that x doesn't delete too many edges
-    if x > g.size():
-        raise ValueError("x must be less than the number of edges in the K-Tree with `n` nodes")
-    
-    for i in range(x):
-        randomEdge = g.random_edge()
-        g.delete_edge(randomEdge)
+    # This formula calculates how many edges are in `k`-tree with `n` nodes
+    if x > ((n*(n - 1)) / 2) + (n - k - 1) * k:
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: The initial clique has k+1 nodes, not n.

File: src/sage/graphs/generators/random.py [hunk 25]
```diff
@@ -1490,25 +1490,23 @@ def RandomKTree(n, k, seed=None):
...
+    A (random) `k`-tree has treewidth `k`::
 
         sage: G = graphs.RandomKTree(50, 5)
         sage: G.treewidth()
         5
         sage: G.show() # not tested 
     """    
     if n < k + 1:
-        raise ValueError("n must be greater than k + 1")
+        raise ValueError("n must be greater than `k + 1`")
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: you don't need the ` around k + 1.

File: src/sage/graphs/generators/random.py (hunk 24 above)
Comment (damianbasso) [reply src/sage/graphs/generators/random.py]: My mistake, fixed

File: src/sage/graphs/generators/random.py [hunk 26]
```diff
@@ -1460,6 +1460,124 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        5
+    
+    EXAMPLES:
+
+    A (random) `k`-tree has treewidth `k`::
+
+        sage: G = graphs.RandomKTree(50, 5)
+        sage: G.treewidth()
+        5
+        sage: G.show() # not tested
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: 2 spaces before `# not tested`

File: src/sage/graphs/generators/random.py [hunk 27]
```diff
@@ -1460,6 +1460,124 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        230
+        sage: g.treewidth()
+        5
+
+    EXAMPLES:
+
+        sage: G = graphs.RandomPartialKTree(50,5,5)
+        sage: G.treewidth()
+        5
+        sage: G.show()
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: as in method `RandomKTree`,  add `# not tested` with 2 spaces before `#`.

File: src/sage/graphs/generators/random.py [hunk 28]
```diff
@@ -1460,6 +1460,124 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+    """
+    if seed is not None:
+        set_random_seed(seed)
+
+    g = RandomKTree(n,k,seed)
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: coding style: add spaces between parameters: `g = RandomKTree(n, k, seed)`

**Review by dcoudert — CHANGES_REQUESTED** [review]
some details.

**Review by dcoudert — APPROVED** [review]
LGTM.

File: src/sage/graphs/generators/random.py [hunk 29]
```diff
@@ -1460,6 +1460,124 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+
+        sage: g=graphs.RandomPartialKTree(50,5,5)
+        sage: g.order()
+        50
+        sage: g.size()
+        230
+        sage: g.treewidth()
+        5
+
+    EXAMPLES:
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: Should be `EXAMPLES::`

**Review by dcoudert — CHANGES_REQUESTED** [review]

**Comment by dcoudert** [comment]
You should certainly rebase the branch on the last beta to launch all the tests.

File: src/sage/graphs/generators/random.py [hunk 30]
```diff
@@ -1459,6 +1459,122 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        5
+
+    EXAMPLES::
+
+        sage: G = graphs.RandomKTree(50, 5)
+        sage: G.treewidth()
+        5
+        sage: G.show()  # not tested
+    """
+    if n < k + 1:
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py:1526]: Here you could add a test to ensure that both n and k are at non negative. 

Actually, we can currently call the generator with `n=5` and `k =0`. What's the expected behavior ? If it's a graph of order `n` without edges, then all vertices should be added to the graph by using `g = Graph(n, name=f"Random {k}-tree of order {n}")`
When `k=1`, what we get is a random tree, right ?

File: src/sage/graphs/generators/random.py [hunk 31]
```diff
@@ -1459,6 +1459,122 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        sage: G.show()  # not tested
+    """
+    if seed is not None:
+        set_random_seed(seed)
+
+    g = RandomKTree(n, k, seed)
+
+    # Check that x doesn't delete too many edges
+    # This formula calculates how many edges are in `k`-tree with `n` nodes
+    if x > (k ^ 2 + k) / 2 + (n - k - 1) * k:
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: In case of equality, the graph has no edges. May be we can avoid the call to `RandomKTree` in this case ?

File: src/sage/graphs/generators/random.py [hunk 32]
```diff
@@ -1459,6 +1459,122 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        raise ValueError("x must be less than the number of edges in the `k`-tree with `n` nodes")
+
+    from sage.misc.prandom import shuffle
+
+    edges = list(g.edges())
+    # Deletes x random edges from the graph
+    shuffle(edges)
+    g.delete_edges(edges[:x])
+    return g
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py:1648]: You may adapt the name of the graph before returning it.

**Review by dcoudert — CHANGES_REQUESTED** [review]
Some improvements to ensure that the method cannot be called with wrong parameters.

File: src/sage/graphs/generators/random.py (hunk 30 above)
Comment (damianbasso) [reply src/sage/graphs/generators/random.py:1526]: You're correct, expected behaviour would be a graph order `n` without edges. I've added appropriate handling on the latest commit. 

When `k=1`, the result is a random tree. The current implementation does do that successfully, but I suppose a case could be added to call `RandomTree` directly if using their random algorithm is preferable

File: src/sage/graphs/generators/random.py [hunk 33]
```diff
@@ -1556,22 +1567,45 @@ def RandomPartialKTree(n, k, x, seed=None):
...
 
     # Check that x doesn't delete too many edges
-    # This formula calculates how many edges are in `k`-tree with `n` nodes
-    if x > (k ^ 2 + k) / 2 + (n - k - 1) * k:
+    if x > edgesInKTree:
         raise ValueError("x must be less than the number of edges in the `k`-tree with `n` nodes")
 
+    # The graph will have no edges
+    if x == edgesInKTree:
+        g = Graph(0, name=f"Random Partial {k}-tree of order 0")
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: Is this the expected behavior ? For me the graph has `n` vertices, not 0. Indeed, when you remove edges you may obtain isolated vertices, but these vertices remain in the graph.

Comment (damianbasso) [reply src/sage/graphs/generators/random.py]: My mistake, you are correct

File: src/sage/graphs/generators/random.py [hunk 34]
```diff
@@ -1459,6 +1459,156 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+    """
+    if n < 0:
+        raise ValueError("n must not be negative")
+
+    if k < 0:
+        raise ValueError("k must not be negative")
+
+    # A graph with treewidth 0 has no edges
+    if k == 0:
+        g = Graph(n, name=f"Random 0-tree of order {n}")
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: when setting the name of a graph, you should not specify the number of nodes. Currently, we get:
```py
sage: G = graphs.RandomKTree(50, 5)
sage: G
Random 5-tree of order 50: Graph on 50 vertices
sage: g=graphs.RandomPartialKTree(50,5,2)
sage: g
Random Partial 5-tree of order 50: Graph on 50 vertices
```

**Review by dcoudert — CHANGES_REQUESTED** [review]

**Review by dcoudert — APPROVED** [review]
LGTM.

File: src/sage/graphs/generators/random.py [hunk 35]
```diff
@@ -1559,6 +1574,30 @@ def RandomPartialKTree(n, k, x, seed=None):
...
         5
+        sage: graphs.RandomPartialKTree(-5, 5, 2)
+        Traceback (most recent call last):
+        ...
+        ValueError: n must not be negative
+        sage: graphs.RandomPartialKTree(5, -5, 2)
+        Traceback (most recent call last):
+        ...
+        ValueError: k must not be negative
+        sage: G = graphs.RandomPartialKTree(2, 5, 2, 1)
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: only 3 parameters

**Review by dcoudert — CHANGES_REQUESTED** [review]

**Comment by dcoudert** [comment]
I don't understand the last 2 commits. Is it due to the last beta release ?

You should certainly rebase the branch on the last beta.

File: src/sage/graphs/generators/random.py [hunk 36]
```diff
@@ -1459,6 +1459,195 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        Traceback (most recent call last):
+        ...
+        ValueError: x must be less than the number of edges in the `k`-tree with `n` nodes
+        sage: G = graphs.RandomPartialKTree(50, 0, 0)
+        sage: G.treewidth()
+        0
+        sage: G = graphs.RandomPartialKTree(5, 2, 7)
+        sage: G.treewidth()
+        0
+        sage :g.size()
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: `sage :g.size()` -> `sage: g.size()`

**Review by dcoudert — CHANGES_REQUESTED** [review]

File: src/sage/graphs/generators/random.py [hunk 37]
```diff
@@ -1459,6 +1459,195 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
...
+        Traceback (most recent call last):
+        ...
+        ValueError: x must be less than the number of edges in the `k`-tree with `n` nodes
+        sage: G = graphs.RandomPartialKTree(50, 0, 0)
+        sage: G.treewidth()
+        0
+        sage: G = graphs.RandomPartialKTree(5, 2, 7)
+        sage: G.treewidth()
+        0
+        sage : g.size()
```
Comment (dcoudert) [inline src/sage/graphs/generators/random.py]: no, there should be no space between `sage` and `:` ;)

We are almost done !

**Review by dcoudert — CHANGES_REQUESTED** [review]

File: src/sage/graphs/generators/random.py (hunk 37 above)
Comment (damianbasso) [reply src/sage/graphs/generators/random.py]: ah my mistake. Thanks

Comment (dcoudert) [reply src/sage/graphs/generators/random.py]: I think we both need an extra coffee today. Please change to `sage: G.size()`.

Currently the test is with `g.size()` and `g` is another graph with many edges...

**Review by dcoudert — APPROVED** [review]
I have no more comments. Thanks.

LGTM.
